### PR TITLE
Oil Pump Optimization

### PR DESCRIPTION
**Pull request by coco0325**
## Description
Improve oil pump on idle

## Changes
Check the buckets in input slot first then check oil supplies. 
If there'are buckets and no supplies, move the buckets to the output slot.
So there won't be too many getChunk() executed when idle.

## Related Issues
Oil pump lags.

### CODE REVIEW

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 1]
```diff
@@ -97,24 +97,29 @@ protected void tick(Block b) {
...
-			Chunk chunk = b.getChunk();
-			int supplies = OreGenSystem.getSupplies(oil, chunk, false);
-			
-			if (supplies > 0) {
-				for (int slot : getInputSlots()) {
-					if (SlimefunManager.isItemSimilar(inv.getItemInSlot(slot), new ItemStack(Material.BUCKET), true)) {
+			for (int slot : getInputSlots()) {
+				if (SlimefunManager.isItemSimilar(inv.getItemInSlot(slot), new ItemStack(Material.BUCKET), true)) {
+					OreGenResource oil = OreGenSystem.getResource("Oil");
+					Chunk chunk = b.getChunk();
```
Comment (WalshyDev) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: Doing this multiple times does not seem like a good idea.

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 2]
```diff
@@ -97,24 +97,29 @@ protected void tick(Block b) {
...
+					int supplies = OreGenSystem.getSupplies(oil, chunk, false);
```
Comment (WalshyDev) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: This already takes an average of 3ms, why not fire it multiple times?

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 3]
```diff
@@ -97,24 +97,29 @@ protected void tick(Block b) {
...
+					if(supplies > 0){
 						MachineRecipe r = new MachineRecipe(26, new ItemStack[0], new ItemStack[] {SlimefunItems.BUCKET_OF_OIL});
-						
+
 						if (!inv.fits(SlimefunItems.BUCKET_OF_OIL, getOutputSlots())) {
 							return;
 						}
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: This check should be moved before the Chunk-call too.

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 4]
```diff
@@ -97,24 +97,29 @@ protected void tick(Block b) {
...
-						
+
 						inv.replaceExistingItem(slot, InvUtils.decreaseItem(inv.getItemInSlot(slot), 1));
 						processing.put(b, r);
 						progress.put(b, r.getTicks());
 						OreGenSystem.setSupplies(oil, chunk, supplies - 1);
 						break;
+					}else{
```
Comment (WalshyDev) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: Formatting

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 5]
```diff
@@ -97,24 +97,29 @@ protected void tick(Block b) {
...
+						ItemStack item = BlockStorage.getInventory(b).getItemInSlot(slot);
+						ItemStack[] abucket = new ItemStack[1];
+						abucket[0]=item.clone();
```
Comment (WalshyDev) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: Formatting and naming. Why is this an array hough?

**Comment by J3fftw1** [comment]
maybe only run the .getSupplies when there is something inside the oilpump

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 6]
```diff
@@ -97,24 +97,29 @@ protected void tick(Block b) {
...
+						BlockStorage.getInventory(b).replaceExistingItem(slot, null);
+						pushItems(b, abucket);
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: This code makes little sense...

First off: Why are you creating completely unncessary arrays like that?

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 7]
```diff
@@ -97,24 +97,29 @@ protected void tick(Block b) {
...
-			
-			if (supplies > 0) {
-				for (int slot : getInputSlots()) {
-					if (SlimefunManager.isItemSimilar(inv.getItemInSlot(slot), new ItemStack(Material.BUCKET), true)) {
+			for (int slot : getInputSlots()) {
+				if (SlimefunManager.isItemSimilar(inv.getItemInSlot(slot), new ItemStack(Material.BUCKET), true)) {
+					OreGenResource oil = OreGenSystem.getResource("Oil");
+					Chunk chunk = b.getChunk();
+					int supplies = OreGenSystem.getSupplies(oil, chunk, false);
+					if(supplies > 0){
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: Please follow the standard formatting here, a space should be added before and after the normal brackets.

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 8]
```diff
@@ -97,25 +97,28 @@ protected void tick(Block b) {
...
 						processing.put(b, r);
 						progress.put(b, r.getTicks());
 						OreGenSystem.setSupplies(oil, chunk, supplies - 1);
 						break;
 					}
+					else {
+						ItemStack item = BlockStorage.getInventory(b).getItemInSlot(slot).clone();
+						BlockStorage.getInventory(b).replaceExistingItem(slot, null);
+						pushItems(b, item);
+					}
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: This code still doesn't make much sense...
Why are you putting the bucket into the machine twice...?

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 9]
```diff
@@ -97,25 +97,28 @@ protected void tick(Block b) {
...
-					if (SlimefunManager.isItemSimilar(inv.getItemInSlot(slot), new ItemStack(Material.BUCKET), true)) {
+			if (!inv.fits(SlimefunItems.BUCKET_OF_OIL, getOutputSlots())) {
+				return;
+			}
+			for (int slot : getInputSlots()) {
+				if (SlimefunManager.isItemSimilar(inv.getItemInSlot(slot), new ItemStack(Material.BUCKET), true)) {
+					OreGenResource oil = OreGenSystem.getResource("Oil");
+					Chunk chunk = b.getChunk();
+					int supplies = OreGenSystem.getSupplies(oil, chunk, false);
+					if(supplies > 0) {
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: Please add a space after the "if" here

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java (hunk 1 above)
Comment (coco0325) [reply src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: Which?

Comment (TheBusyBiscuit) [reply src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: getChunk()

Comment (WalshyDev) [reply src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: All of this. You don't need to get the resource, get the chunk and get the supplies every input slot.

Comment (TheBusyBiscuit) [reply src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: There should be a break; at the end, then it won't be a problem since it will never run more than once anyway.

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 10]
```diff
@@ -97,28 +97,31 @@ protected void tick(Block b) {
...
 						processing.put(b, r);
 						progress.put(b, r.getTicks());
 						OreGenSystem.setSupplies(oil, chunk, supplies - 1);
 						break;
 					}
+					else {
+						ItemStack item = BlockStorage.getInventory(b).getItemInSlot(slot).clone();
+						BlockStorage.getInventory(b).replaceExistingItem(slot, null);
+						pushItems(b, item);
+						break;
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: For the third time: Why did you add this...?

Comment (coco0325) [reply src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: Moving buckets from input to output slot.

Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: You are breaking in both cases of the if/else clause, so you might as well put one single break; after both.

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 11]
```diff
@@ -97,28 +97,31 @@ protected void tick(Block b) {
...
-			OreGenResource oil = OreGenSystem.getResource("Oil");
-			Chunk chunk = b.getChunk();
-			int supplies = OreGenSystem.getSupplies(oil, chunk, false);
-			
-			if (supplies > 0) {
-				for (int slot : getInputSlots()) {
-					if (SlimefunManager.isItemSimilar(inv.getItemInSlot(slot), new ItemStack(Material.BUCKET), true)) {
+			if (!inv.fits(SlimefunItems.BUCKET_OF_OIL, getOutputSlots())) {
+				return;
+			}
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: You could refactor this to simply be

else if (inv.fits(SlimefunItems.BUCKET_OF_OIL, getOutputSlots())) {
// all other code here
}

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java [hunk 12]
```diff
@@ -96,29 +96,28 @@ protected void tick(Block b) {
...
 		}
-		else {
-			OreGenResource oil = OreGenSystem.getResource("Oil");
-			Chunk chunk = b.getChunk();
-			int supplies = OreGenSystem.getSupplies(oil, chunk, false);
-			
-			if (supplies > 0) {
-				for (int slot : getInputSlots()) {
-					if (SlimefunManager.isItemSimilar(inv.getItemInSlot(slot), new ItemStack(Material.BUCKET), true)) {
+		else if(inv.fits(SlimefunItems.BUCKET_OF_OIL, getOutputSlots())) {
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/machines/electric/geo/OilPump.java]: Please add a space after the "if".
